src/UNIV2LPOracle.sol

```diff
@@ -264,14 +294,67 @@ contract UNIV2LPOracle {
         );
     }
 
-    function poke() external stoppable {
-        require(pass(), "UNIV2LPOracle/not-passed");
-        uint128 val = seek();
-        require(val != 0, "UNIV2LPOracle/invalid-price");
-        cur = nxt;
-        nxt = Feed(val, 1);
-        zzz = uint64(block.timestamp);
-        emit Value(cur.val, nxt.val);
+    function poke() external {
+
+        // Ensure a single SLOAD while avoiding solc's excessive bitmasking bureaucracy.
+        uint256 _zph;
+        uint256 _hop;
+        {
+            uint256 _stopped;  // block-scoping _stopped here saves a little gas
+            assembly {
+                _zph     := sload(1)
+                _stopped := and(_zph,         0xff    )
+                _hop     := and(shr(8, _zph), 0xffffff)
+                _zph     := shr(32, _zph)
+            }
+
+            // When stopped, values are set to zero and should remain such; thus, disallow updating in that case.
+            require(_stopped == 0, "UNIV2LPOracle/is-stopped");
+        }
+
+        // Equivalent to requiring that pass() returns true;
+        // code has been duplicated to reduce gas costs.
+        require(block.timestamp >= _zph, "UNIV2LPOracle/not-passed");
+
+        uint128 _val = seek();
+        require(_val != 0, "UNIV2LPOracle/invalid-price");
+        Feed memory _cur = nxt;  // This memory value is used to save an SLOAD later.
+        cur = _cur;
+        nxt = Feed(_val, 1);
+
+        // The below is equivalent to:
+        //
+        //    zph = block.timestamp + hop
+        //
+        // but ensures no extra SLOADs are performed.
+        //
+        // Even if _hop = (2^24 - 1), the maximum possible value, add(timestamp(), _hop)
+        // will not overflow (even a 224 bit value) for a very long time.
+        //
+        // Also, we know stopped was zero, so there is no need to account for it explicitly here.
+        assembly {
+            sstore(
+                1,
```

Comment:
These are very specific to the memory slots they're in. I am assuming that if these variables get shifted in another PR this stuff will break the tests, but it might be worthwhile to verify that's the case.

Reply:
Oh yes the tests will break, even if the values just aren't written correctly in these slots, since we read `zph` and `hop` via their public accessors in the `poke` test.